#  Importing Pakages

In [1]:
import numpy as np
import pandas as pd
import pandas_profiling as pp

# Loading Data

In [35]:
#Loading data
data1=pd.read_csv('C:\Wine-Quality-Prdiction\Wine-Quality-Prdiction\Data Set\winequality_red.csv')
data2=pd.read_csv('C:\Wine-Quality-Prdiction\Wine-Quality-Prdiction\Data Set\winequality_white.csv')

# Merging dataframes

In [36]:
#Number of rows in first dataframe
length1=data1.shape[0]
length1

1599

In [4]:
#Number of rows in second datafram
length2=data2.shape[0]
length2


4898

In [5]:
#Additon of length of both rows
length=length1+length2
length

6497

In [6]:
#Addition of row in first dataframe
updated_index=np.arange(length2,length)
data1['updated_index']=updated_index
updated_index
data1

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                  34.0  0.99780  3.51       0.56   
1                    25.0                  67.0  0.99680  3.20       0.68   
2                    15.0                  54.0  0.99700  3.26       0.65   
3                    17.0                  60.0  0.99800  3.16       0.58   
4                    11.0                  34.0  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  updated_index  
0         9.4        5           4898  
1         9.8        5           4899  
2         9.8        5           4900  
3         9.8        6           4901  
4         9.4        5           4902  
...       ...      ...            ...  
1594     10.5        5           6492  
1595     11.2        6           6493  
1596     11.0        6           6494  
1597     10.2        5           6495  
1598     11.0        6           6496  

[1599 rows x 13 columns]

In [7]:
#Saving previous index with name index in dataframe1
index=data1.index
data1['old_index']=index

#Concatinating first and second dataframe 
data1.set_index('updated_index', inplace = True)
data = pd.concat([data2,data1])
data

C:\Users\Shivam\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  import sys


alcohol  chlorides  citric acid  density  fixed acidity  \
0         8.8      0.045         0.36  1.00100            7.0   
1         9.5      0.049         0.34  0.99400            6.3   
2        10.1      0.050         0.40  0.99510            8.1   
3         9.9      0.058         0.32  0.99560            7.2   
4         9.9      0.058         0.32  0.99560            7.2   
...       ...        ...          ...      ...            ...   
6492     10.5      0.090         0.08  0.99490            6.2   
6493     11.2      0.062         0.10  0.99512            5.9   
6494     11.0      0.076         0.13  0.99574            6.3   
6495     10.2      0.075         0.12  0.99547            5.9   
6496     11.0      0.067         0.47  0.99549            6.0   

      free sulfur dioxide  old_index    pH  quality  residual sugar  \
0                    45.0        NaN  3.00        6            20.7   
1                    14.0        NaN  3.30        6             1.6   
2                    30.0        NaN  3.26        6             6.9   
3                    47.0        NaN  3.19        6             8.5   
4                    47.0        NaN  3.19        6             8.5   
...                   ...        ...   ...      ...             ...   
6492                 32.0     1594.0  3.45        5             2.0   
6493                 39.0     1595.0  3.52        6             2.2   
6494                 29.0     1596.0  3.42        6             2.3   
6495                 32.0     1597.0  3.57        5             2.0   
6496                 18.0     1598.0  3.39        6             3.6   

      sulphates  total sulfur dioxide  volatile acidity  
0          0.45                 170.0             0.270  
1          0.49                 132.0             0.300  
2          0.44                  97.0             0.280  
3          0.40                 186.0             0.230  
4          0.40                 186.0             0.230  
...         ...                   ...               ...  
6492       0.58                  44.0             0.600  
6493       0.76                  51.0             0.550  
6494       0.75                  40.0             0.510  
6495       0.71                  44.0             0.645  
6496       0.66                  42.0             0.310  

[6497 rows x 13 columns]

# Describing data

In [8]:
#Top 5 rows
data.head(5)

alcohol  chlorides  citric acid  density  fixed acidity  \
0      8.8      0.045         0.36   1.0010            7.0   
1      9.5      0.049         0.34   0.9940            6.3   
2     10.1      0.050         0.40   0.9951            8.1   
3      9.9      0.058         0.32   0.9956            7.2   
4      9.9      0.058         0.32   0.9956            7.2   

   free sulfur dioxide  old_index    pH  quality  residual sugar  sulphates  \
0                 45.0        NaN  3.00        6            20.7       0.45   
1                 14.0        NaN  3.30        6             1.6       0.49   
2                 30.0        NaN  3.26        6             6.9       0.44   
3                 47.0        NaN  3.19        6             8.5       0.40   
4                 47.0        NaN  3.19        6             8.5       0.40   

   total sulfur dioxide  volatile acidity  
0                 170.0              0.27  
1                 132.0              0.30  
2                  97.0              0.28  
3                 186.0              0.23  
4                 186.0              0.23

In [9]:
#Last 5 rows
data.tail(5)

alcohol  chlorides  citric acid  density  fixed acidity  \
6492     10.5      0.090         0.08  0.99490            6.2   
6493     11.2      0.062         0.10  0.99512            5.9   
6494     11.0      0.076         0.13  0.99574            6.3   
6495     10.2      0.075         0.12  0.99547            5.9   
6496     11.0      0.067         0.47  0.99549            6.0   

      free sulfur dioxide  old_index    pH  quality  residual sugar  \
6492                 32.0     1594.0  3.45        5             2.0   
6493                 39.0     1595.0  3.52        6             2.2   
6494                 29.0     1596.0  3.42        6             2.3   
6495                 32.0     1597.0  3.57        5             2.0   
6496                 18.0     1598.0  3.39        6             3.6   

      sulphates  total sulfur dioxide  volatile acidity  
6492       0.58                  44.0             0.600  
6493       0.76                  51.0             0.550  
6494       0.75                  40.0             0.510  
6495       0.71                  44.0             0.645  
6496       0.66                  42.0             0.310

In [10]:
#Data Size
data.size

84461

In [11]:
#Shape of dataset
print(data.shape)
rows,column=data.shape
print('Rows are {0} \nColumns are {1}'.format(rows,column))

(6497, 13)
Rows are 6497 
Columns are 13


In [12]:
#DataSet Index
data.index

Int64Index([   0,    1,    2,    3,    4,    5,    6,    7,    8,    9,
            ...
            6487, 6488, 6489, 6490, 6491, 6492, 6493, 6494, 6495, 6496],
           dtype='int64', length=6497)

In [13]:
#Column names
data.columns

Index(['alcohol', 'chlorides', 'citric acid', 'density', 'fixed acidity',
       'free sulfur dioxide', 'old_index', 'pH', 'quality', 'residual sugar',
       'sulphates', 'total sulfur dioxide', 'volatile acidity'],
      dtype='object')

In [14]:
#Information about data 
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6497 entries, 0 to 6496
Data columns (total 13 columns):
alcohol                 6497 non-null float64
chlorides               6497 non-null float64
citric acid             6497 non-null float64
density                 6497 non-null float64
fixed acidity           6497 non-null float64
free sulfur dioxide     6497 non-null float64
old_index               1599 non-null float64
pH                      6497 non-null float64
quality                 6497 non-null int64
residual sugar          6497 non-null float64
sulphates               6497 non-null float64
total sulfur dioxide    6497 non-null float64
volatile acidity        6497 non-null float64
dtypes: float64(12), int64(1)
memory usage: 710.6 KB


In [15]:
#Data Description
data.describe()

alcohol    chlorides  citric acid      density  fixed acidity  \
count  6497.000000  6497.000000  6497.000000  6497.000000    6497.000000   
mean     10.491801     0.056034     0.318633     0.994697       7.215307   
std       1.192712     0.035034     0.145318     0.002999       1.296434   
min       8.000000     0.009000     0.000000     0.987110       3.800000   
25%       9.500000     0.038000     0.250000     0.992340       6.400000   
50%      10.300000     0.047000     0.310000     0.994890       7.000000   
75%      11.300000     0.065000     0.390000     0.996990       7.700000   
max      14.900000     0.611000     1.660000     1.038980      15.900000   

       free sulfur dioxide    old_index           pH      quality  \
count          6497.000000  1599.000000  6497.000000  6497.000000   
mean             30.525319   799.000000     3.218501     5.818378   
std              17.749400   461.735855     0.160787     0.873255   
min               1.000000     0.000000     2.720000     3.000000   
25%              17.000000   399.500000     3.110000     5.000000   
50%              29.000000   799.000000     3.210000     6.000000   
75%              41.000000  1198.500000     3.320000     6.000000   
max             289.000000  1598.000000     4.010000     9.000000   

       residual sugar    sulphates  total sulfur dioxide  volatile acidity  
count     6497.000000  6497.000000           6497.000000       6497.000000  
mean         5.443235     0.531268            115.744574          0.339666  
std          4.757804     0.148806             56.521855          0.164636  
min          0.600000     0.220000              6.000000          0.080000  
25%          1.800000     0.430000             77.000000          0.230000  
50%          3.000000     0.510000            118.000000          0.290000  
75%          8.100000     0.600000            156.000000          0.400000  
max         65.800000     2.000000            440.000000          1.580000

In [16]:
#Check datatype of each attribute
data.dtypes

alcohol                 float64
chlorides               float64
citric acid             float64
density                 float64
fixed acidity           float64
free sulfur dioxide     float64
old_index               float64
pH                      float64
quality                   int64
residual sugar          float64
sulphates               float64
total sulfur dioxide    float64
volatile acidity        float64
dtype: object

In [17]:
#Types of datatype in our dataframes
data.get_dtype_counts()


C:\Users\Shivam\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: `get_dtype_counts` has been deprecated and will be removed in a future version. For DataFrames use `.dtypes.value_counts()
  


float64    12
int64       1
dtype: int64

In [18]:
data.count()

alcohol                 6497
chlorides               6497
citric acid             6497
density                 6497
fixed acidity           6497
free sulfur dioxide     6497
old_index               1599
pH                      6497
quality                 6497
residual sugar          6497
sulphates               6497
total sulfur dioxide    6497
volatile acidity        6497
dtype: int64

# Remove Insignificant columns


In [19]:
data

alcohol  chlorides  citric acid  density  fixed acidity  \
0         8.8      0.045         0.36  1.00100            7.0   
1         9.5      0.049         0.34  0.99400            6.3   
2        10.1      0.050         0.40  0.99510            8.1   
3         9.9      0.058         0.32  0.99560            7.2   
4         9.9      0.058         0.32  0.99560            7.2   
...       ...        ...          ...      ...            ...   
6492     10.5      0.090         0.08  0.99490            6.2   
6493     11.2      0.062         0.10  0.99512            5.9   
6494     11.0      0.076         0.13  0.99574            6.3   
6495     10.2      0.075         0.12  0.99547            5.9   
6496     11.0      0.067         0.47  0.99549            6.0   

      free sulfur dioxide  old_index    pH  quality  residual sugar  \
0                    45.0        NaN  3.00        6            20.7   
1                    14.0        NaN  3.30        6             1.6   
2                    30.0        NaN  3.26        6             6.9   
3                    47.0        NaN  3.19        6             8.5   
4                    47.0        NaN  3.19        6             8.5   
...                   ...        ...   ...      ...             ...   
6492                 32.0     1594.0  3.45        5             2.0   
6493                 39.0     1595.0  3.52        6             2.2   
6494                 29.0     1596.0  3.42        6             2.3   
6495                 32.0     1597.0  3.57        5             2.0   
6496                 18.0     1598.0  3.39        6             3.6   

      sulphates  total sulfur dioxide  volatile acidity  
0          0.45                 170.0             0.270  
1          0.49                 132.0             0.300  
2          0.44                  97.0             0.280  
3          0.40                 186.0             0.230  
4          0.40                 186.0             0.230  
...         ...                   ...               ...  
6492       0.58                  44.0             0.600  
6493       0.76                  51.0             0.550  
6494       0.75                  40.0             0.510  
6495       0.71                  44.0             0.645  
6496       0.66                  42.0             0.310  

[6497 rows x 13 columns]

We have no insignificant columns in our data set

# Check Missing value for each columns

In [20]:
data.isna().sum()

alcohol                    0
chlorides                  0
citric acid                0
density                    0
fixed acidity              0
free sulfur dioxide        0
old_index               4898
pH                         0
quality                    0
residual sugar             0
sulphates                  0
total sulfur dioxide       0
volatile acidity           0
dtype: int64

We have 4898 nan value in index column.

# Remove rows with missing values

We will convert all null values of old_index column into -1 so that we can set a difference between both dataframes

In [21]:
#Filling all nan value with -1
data.old_index.fillna(-1,inplace=True)
data

alcohol  chlorides  citric acid  density  fixed acidity  \
0         8.8      0.045         0.36  1.00100            7.0   
1         9.5      0.049         0.34  0.99400            6.3   
2        10.1      0.050         0.40  0.99510            8.1   
3         9.9      0.058         0.32  0.99560            7.2   
4         9.9      0.058         0.32  0.99560            7.2   
...       ...        ...          ...      ...            ...   
6492     10.5      0.090         0.08  0.99490            6.2   
6493     11.2      0.062         0.10  0.99512            5.9   
6494     11.0      0.076         0.13  0.99574            6.3   
6495     10.2      0.075         0.12  0.99547            5.9   
6496     11.0      0.067         0.47  0.99549            6.0   

      free sulfur dioxide  old_index    pH  quality  residual sugar  \
0                    45.0       -1.0  3.00        6            20.7   
1                    14.0       -1.0  3.30        6             1.6   
2                    30.0       -1.0  3.26        6             6.9   
3                    47.0       -1.0  3.19        6             8.5   
4                    47.0       -1.0  3.19        6             8.5   
...                   ...        ...   ...      ...             ...   
6492                 32.0     1594.0  3.45        5             2.0   
6493                 39.0     1595.0  3.52        6             2.2   
6494                 29.0     1596.0  3.42        6             2.3   
6495                 32.0     1597.0  3.57        5             2.0   
6496                 18.0     1598.0  3.39        6             3.6   

      sulphates  total sulfur dioxide  volatile acidity  
0          0.45                 170.0             0.270  
1          0.49                 132.0             0.300  
2          0.44                  97.0             0.280  
3          0.40                 186.0             0.230  
4          0.40                 186.0             0.230  
...         ...                   ...               ...  
6492       0.58                  44.0             0.600  
6493       0.76                  51.0             0.550  
6494       0.75                  40.0             0.510  
6495       0.71                  44.0             0.645  
6496       0.66                  42.0             0.310  

[6497 rows x 13 columns]

In [22]:
data.isna().sum()

alcohol                 0
chlorides               0
citric acid             0
density                 0
fixed acidity           0
free sulfur dioxide     0
old_index               0
pH                      0
quality                 0
residual sugar          0
sulphates               0
total sulfur dioxide    0
volatile acidity        0
dtype: int64

Now we have clear data with no nan value

In [23]:
#printing duplicate rows
duplicate=(data.duplicated())
data[duplicate]

alcohol  chlorides  citric acid  density  fixed acidity  \
4      9.900000      0.058         0.32  0.99560            7.2   
5     10.100000      0.050         0.40  0.99510            8.1   
7      8.800000      0.045         0.36  1.00100            7.0   
8      9.500000      0.049         0.34  0.99400            6.3   
20    12.800000      0.029         0.48  0.98920            6.2   
...         ...        ...          ...      ...            ...   
4828   9.200000      0.042         0.35  0.99670            6.4   
4850  11.100000      0.048         0.35  0.99146            7.0   
4851   9.600000      0.055         0.44  0.99488            6.4   
4856   9.000000      0.058         0.39  0.99755            7.1   
4880   9.533333      0.046         0.40  0.99494            6.6   

      free sulfur dioxide  old_index    pH  quality  residual sugar  \
4                    47.0       -1.0  3.19        6             8.5   
5                    30.0       -1.0  3.26        6             6.9   
7                    45.0       -1.0  3.00        6            20.7   
8                    14.0       -1.0  3.30        6             1.6   
20                   29.0       -1.0  3.33        8             1.2   
...                   ...        ...   ...      ...             ...   
4828                 54.0       -1.0  3.23        5            10.3   
4850                 67.0       -1.0  3.05        6             2.5   
4851                 52.0       -1.0  3.10        5             8.9   
4856                 26.0       -1.0  2.90        6            13.7   
4880                 68.0       -1.0  3.15        6             8.1   

      sulphates  total sulfur dioxide  volatile acidity  
4          0.40                 186.0              0.23  
5          0.44                  97.0              0.28  
7          0.45                 170.0              0.27  
8          0.49                 132.0              0.30  
20         0.39                  75.0              0.66  
...         ...                   ...               ...  
4828       0.47                 140.0              0.23  
4850       0.56                 161.0              0.36  
4851       0.48                 164.0              0.33  
4856       0.46                 172.0              0.23  
4880       0.50                 170.0              0.34  

[937 rows x 13 columns]

The above printed rows are duplicate rows

In [24]:
data[duplicate].shape

(937, 13)

There are 937 duplicate rows

Removing duplicate rows

In [25]:
data.drop_duplicates(inplace=True)

In [26]:
#New shape of data
data.shape

(5560, 13)

Now there are 5560 rows and 13 columns in our data frame

In [27]:
data.size

72280

Size of updated data frame is 72280

# Formatting your Data – making data types compaitable with other data types.

In [28]:
#Check datatype of each attribute
data.dtypes

alcohol                 float64
chlorides               float64
citric acid             float64
density                 float64
fixed acidity           float64
free sulfur dioxide     float64
old_index               float64
pH                      float64
quality                   int64
residual sugar          float64
sulphates               float64
total sulfur dioxide    float64
volatile acidity        float64
dtype: object

In [29]:
#Types of datatype in our dataframes
data.get_dtype_counts()


C:\Users\Shivam\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: `get_dtype_counts` has been deprecated and will be removed in a future version. For DataFrames use `.dtypes.value_counts()
  


float64    12
int64       1
dtype: int64

Since we have compatible data so we do not need to formate our data type

Saving clean and merged dataframe in another csv file

In [30]:
#Sorting data through quality and old_index column
list_1=['quality','old_index']
data.sort_values(list_1, axis = 0, ascending = True, 
                 inplace = True, na_position ='last') 
data 


alcohol  chlorides  citric acid  density  fixed acidity  \
251       9.8      0.074         0.21  0.99800            8.5   
253      11.7      0.029         0.44  0.99130            5.8   
294       8.5      0.066         0.38  0.99680            9.1   
445      11.5      0.038         0.32  0.99370            7.1   
740      12.6      0.022         0.40  0.99150            6.9   
...       ...        ...          ...      ...            ...   
774      10.4      0.035         0.45  0.99700            9.1   
820      12.4      0.021         0.29  0.98965            6.6   
827      12.5      0.031         0.36  0.99055            7.4   
876      12.7      0.018         0.34  0.98980            6.9   
1605     12.9      0.032         0.49  0.99030            7.1   

      free sulfur dioxide  old_index    pH  quality  residual sugar  \
251                  41.0       -1.0  3.02        3            16.2   
253                   5.0       -1.0  3.53        3             3.5   
294                  34.0       -1.0  3.23        3             1.6   
445                  16.0       -1.0  3.24        3            11.0   
740                   5.0       -1.0  3.31        3             4.6   
...                   ...        ...   ...      ...             ...   
774                  28.0       -1.0  3.20        9            10.6   
820                  24.0       -1.0  3.41        9             1.6   
827                  27.0       -1.0  3.28        9             2.0   
876                  57.0       -1.0  3.28        9             4.2   
1605                 31.0       -1.0  3.37        9             2.2   

      sulphates  total sulfur dioxide  volatile acidity  
251        0.50                 197.0              0.26  
253        0.43                 109.0              0.24  
294        0.38                 182.0              0.59  
445        0.40                  66.0              0.32  
740        0.37                  19.0              0.39  
...         ...                   ...               ...  
774        0.46                 124.0              0.27  
820        0.61                  85.0              0.36  
827        0.48                 139.0              0.24  
876        0.36                 119.0              0.36  
1605       0.42                 113.0              0.26  

[5560 rows x 13 columns]

In [31]:
data.to_csv('Merged_dataset.csv') 